jvcprojector/projector.py

```diff
@@ -101,50 +104,147 @@ async def disconnect(self) -> None:
 
     async def get_info(self) -> dict[str, str]:
         """Get device info."""
-        assert self._device
-        model = JvcCommand(command.MODEL, True)
-        mac = JvcCommand(command.MAC, True)
-        await self._send([model, mac])
+        if not self._device:
+            raise JvcProjectorError("Must call connect before getting info")
+
+        model = JvcCommand(const.CMD_MODEL, True)
+        # ensure PJ is fully on first
+        if await self.is_on():
+            mac = JvcCommand(const.CMD_LAN_SETUP_MAC_ADDRESS, True)
+            version = JvcCommand(const.CMD_VERSION, True)
+            await self._send([model, mac, version])
+        else:
+            await self._send([model])
 
         if mac.response is None:
-            raise JvcProjectorError("Mac address not available")
+            mac.response = "(unknown)"
 
         if model.response is None:
             model.response = "(unknown)"
 
+        # store to reference in state
         self._model = model.response
         self._mac = mac.response
+        self._version = version.response
 
-        return {"model": self._model, "mac": self._mac}
+        return {const.KEY_MODEL: self._model, const.KEY_MAC: self._mac}
 
     async def get_state(self) -> dict[str, str | None]:
         """Get device state."""
-        assert self._device
-        pwr = JvcCommand(command.POWER, True)
-        inp = JvcCommand(command.INPUT, True)
-        src = JvcCommand(command.SOURCE, True)
-        res = await self._send([pwr, inp, src])
-        return {
-            "power": res[0] or None,
-            "input": res[1] or const.NOSIGNAL,
-            "source": res[2] or const.NOSIGNAL,
-        }
+        if not self._device:
+            raise JvcProjectorError("Must call connect before getting state")
+
+        # Add static values
+        self._dict[const.KEY_MODEL] = self.process_model_code(self._model)
+        self._dict[const.KEY_VERSION] = self.process_version(self._version)
+        self._dict[const.KEY_MAC] = self.process_mac(self._mac)
+
+        async def send_and_update(commands: dict[str, str]) -> None:
+            """Send commands and update the dictionary."""
+            cmd_vals = [JvcCommand(cmd, True) for cmd in commands.values()]
+            res = await self._send(cmd_vals)
+            # discard the command values and zip the keys with the responses
+            for (key, _), value in zip(commands.items(), res):
+                self._dict[key] = value
+
+        # Always get power state
+        await send_and_update({const.KEY_POWER: const.CMD_POWER})
+
+        # If power is on, get additional states
+        if self._dict.get(const.KEY_POWER) == const.ON:
+            await send_and_update(
+                {
+                    const.KEY_INPUT: const.CMD_INPUT,
```

Comment:
Guess the HA integration can be reworked some, but it relies on the `input` key being in the response at all times (even it set to NO_SIGNAL). https://github.com/home-assistant/core/blob/dev/homeassistant/components/jvc_projector/select.py#L32

This was erroring out for me when I tested with your PR in HA.

Reply:
the HA changes will require a PR for a version bump so IMO this wont break anything as I plan to add these features to the integration at the same time